Add class reconstruction support.

ILOVEPIE opened the pull request:
This is a PR for class reconstruction support. It's still a work in progress, but it's working. We need to clean it up probably.

ILOVEPIE committed:
Initial work on class reconstruction support.

ILOVEPIE commented:
@cyberkaida let me know what improvements you want me to make to this rough version.

cyberkaida commented on src/main/java/reva/tools/classes/ClassToolProvider.java:
```diff
@@ -0,0 +1,1517 @@
+                    // Get function count for this namespace
+                    FunctionManager funcMgr = program.getFunctionManager();
+                    int functionCount = 0;
+                    for (Function func : funcMgr.getFunctions(true)) {
```
This is a very slow way to do this (iterating all functions for every class).

You can use the [symbol table to efficiently iterate symbols in a Namespace](https://ghidra.re/ghidra_docs/api/ghidra/program/model/symbol/SymbolTable.html#getSymbols(ghidra.program.model.symbol.Namespace)) then you can check each symbol is a function (Use [SymbolType](https://ghidra.re/ghidra_docs/api/ghidra/program/model/symbol/SymbolType.html#FUNCTION))

Another way is to iterate once and cache the parent name space in a map. But I think use the symbol table.

ILOVEPIE replied:
Yeah, this was just a rough first pass.

cyberkaida commented on src/main/java/reva/tools/classes/ClassToolProvider.java:
```diff
@@ -0,0 +1,1517 @@
+        Map<String, Object> properties = new HashMap<>();
+        properties.put("programPath", SchemaUtil.stringProperty("Path in the Ghidra Project to the program"));
+        properties.put("className", SchemaUtil.stringProperty("Name of the class namespace"));
+        properties.put("includeInherited", SchemaUtil.booleanPropertyWithDefault("Include methods from parent classes", false));
```
This is nice 👍

cyberkaida commented on src/main/java/reva/tools/classes/ClassToolProvider.java:
```diff
@@ -0,0 +1,1517 @@
+    /**
+     * Find a data type by name in all categories
+     */
+    private DataType findDataTypeByName(DataTypeManager dtm, String name) {
```
This would be better in the DataTypeUtils.

cyberkaida commented on src/main/java/reva/tools/classes/ClassToolProvider.java:
```diff
@@ -0,0 +1,1517 @@
+     * Classify method type based on name patterns
+     */
+    private String classifyMethodType(String funcName) {
+        if (funcName.contains("ctor") || funcName.contains("Constructor")) {
```
Also add `init` here and `deinit` to the destructor.

cyberkaida commented on src/main/java/reva/tools/classes/ClassToolProvider.java:
```diff
@@ -0,0 +1,1517 @@
+            structInfo.put("guidance", "Use 'get-structure-info' tool with structureName '" + struct.getName() + "' to see detailed field information.");
+        } else {
+            structInfo.put("hasStructure", false);
+            structInfo.put("structureName", null);
```
Why include a null structure name here?

cyberkaida commented on src/main/java/reva/tools/classes/ClassToolProvider.java:
```diff
@@ -0,0 +1,1517 @@
+    /**
+     * Parse a data type string into a Ghidra DataType
+     */
+    private DataType parseDataType(String typeName, DataTypeManager dtm) {
```
There is already a function like this in the DataTypeUtils

cyberkaida committed:
Update ClassToolProvider to follow MCP v0.11.1 patterns

- Use McpSchema.Tool.builder() pattern instead of constructor
- Add title fields to all 8 class tools for better UX
- Fix address formatting to use AddressUtil.formatAddress()
- Update error message to match integration test expectations
- Remove unused imports and add missing AddressUtil import

This aligns the ClassToolProvider with the new MCP SDK patterns
introduced in the main branch rebase.
